### PR TITLE
Fix incorrect BAD_FORMAT_NOTHROW_ASSERTs in casting

### PR DESCRIPTION
**Pull request by jkotas**

### CODE REVIEW

File: src/coreclr/src/vm/typedesc.cpp [hunk 1]
```diff
@@ -40,8 +40,7 @@ BOOL ParamTypeDesc::Verify() {
 
     _ASSERTE(m_TemplateMT.IsNull() || GetTemplateMethodTableInternal()->SanityCheck());
     _ASSERTE(!GetTypeParam().IsNull());
-    BAD_FORMAT_NOTHROW_ASSERT(GetTypeParam().IsTypeDesc() || !GetTypeParam().AsMethodTable()->IsArray());
```
Comment (jkotas) [inline src/coreclr/src/vm/typedesc.cpp]: `BAD_FORMAT_NOTHROW_ASSERT` fires only when you are running with debugger attached. I have seen this fire for `ref int[]` under debugger. This assert assumed that arrays have ParamTypeDescs.

File: src/coreclr/src/vm/typedesc.cpp [hunk 2]
```diff
@@ -40,8 +40,7 @@ BOOL ParamTypeDesc::Verify() {
...
-    BAD_FORMAT_NOTHROW_ASSERT(CorTypeInfo::IsModifier_NoThrow(GetInternalCorElementType()) ||
```
Comment (jkotas) [inline src/coreclr/src/vm/typedesc.cpp]: Change this to be a hard assert. We should never come here with bogus element type.

File: src/coreclr/src/vm/typedesc.cpp [hunk 3]
```diff
@@ -387,18 +386,12 @@ BOOL TypeDesc::CanCastTo(TypeHandle toTypeHnd, TypeHandlePairList *pVisited)
             case ELEMENT_TYPE_PTR:
                 fCast = TypeDesc::CanCastParam(dac_cast<PTR_ParamTypeDesc>(this)->GetTypeParam(), dac_cast<PTR_ParamTypeDesc>(toTypeDesc)->GetTypeParam(), pVisited);
                 break;
-            case ELEMENT_TYPE_VAR:
-            case ELEMENT_TYPE_MVAR:
-            case ELEMENT_TYPE_FNPTR:
-                fCast = FALSE;
-                break;
             default:
-                BAD_FORMAT_NOTHROW_ASSERT(toKind == ELEMENT_TYPE_TYPEDBYREF || CorTypeInfo::IsPrimitiveType(toKind));
```
Comment (jkotas) [inline src/coreclr/src/vm/typedesc.cpp]: This was unreachable path, so I have refactored it a bit for clarity.

File: src/coreclr/src/vm/typedesc.cpp [hunk 4]
```diff
@@ -387,18 +386,12 @@ BOOL TypeDesc::CanCastTo(TypeHandle toTypeHnd, TypeHandlePairList *pVisited)
...
                 // array cast should have been handled above
                 _ASSERTE(toKind != ELEMENT_TYPE_ARRAY);
```
Comment (VSadov) [inline src/coreclr/src/vm/typedesc.cpp]: These asserts and comment could be removed too. They do not make much sense now. Arrays are no longer handled in this method.

File: src/coreclr/src/vm/typedesc.cpp (hunk 1 above)
Comment (VSadov) [reply src/coreclr/src/vm/typedesc.cpp]: Looks like I missed this one. Thanks!

---
In reply to: [371405798](https://github.com/dotnet/runtime/pull/2237#discussion_r371405798) [](ancestors = 371405798)

**Review by VSadov — APPROVED** [review]
Thanks!

File: src/coreclr/src/vm/typedesc.cpp [hunk 5]
```diff
@@ -387,18 +386,6 @@ BOOL TypeDesc::CanCastTo(TypeHandle toTypeHnd, TypeHandlePairList *pVisited)
             case ELEMENT_TYPE_PTR:
                 fCast = TypeDesc::CanCastParam(dac_cast<PTR_ParamTypeDesc>(this)->GetTypeParam(), dac_cast<PTR_ParamTypeDesc>(toTypeDesc)->GetTypeParam(), pVisited);
```
Comment (VSadov) [inline src/coreclr/src/vm/typedesc.cpp]: Looks like GCC wants a `default`. Even if an empty one.

**Pull request merged**